### ISSUE TITLE
Configure CircuitBreaker threshold using error rate over time window

### ISSUE THREAD

**Issue by armelbd** [jhalterman/failsafe#214]
In the current version, CircuitBreakers switch to the Open state when a given number of failures f occurred over the last e executions:
```Java
CircuitBreaker<Object> breaker = new CircuitBreaker<>()
  .withFailureThreshold(f, e);
```
In my experience, coming up for a number for f and e is hard, especially when the number of executions changes dramatically over time (e.g traffic peak for a web application). I propose to support an error rate over a time window, the API could be something like:
```Java
CircuitBreaker<Object> breaker = new CircuitBreaker<>()
// Open the circuit if we have more than 10% of errors in a rolling 30 seconds window
  .withFailureThreshold(0.1, Duration.ofSeconds(30));
```
I think it would be reasonable to have the error rate refreshed every second and a max window of 1 hour (worst case 28.125 KB memory footprint). Or to provide pre-configured windows if we want a simpler approach:
```Java
CircuitBreaker<Object> breaker = new CircuitBreaker<>()
  .withFailureThresholdOneMinute(0.1)
  .withFailureThresholdFiveMinute(0.05);
```
Note that the last example has 2 different thresholds for 1 and 5 minutes window. It would be nice but I guess having multiple CircuitBreakers chain together would achieve the same result.

Please let me know what you all think, I will be happy to contribute a PR for this feature

**Comment by whiskeysierra** [comment]
Same as #208, isn't it?

**Comment by armelbd** [comment]
Yup, sorry I miss yours

**Comment by armelbd** [comment]
Also implementing this (at least the way I'm thinking about it) would give us rate limiting for almost free (#78)

**Comment by jhalterman** [comment]
I'd be happy to take a PR for this. To simplify the discussion though, I'm going to close this as a duplicate of #208 and take the conversation over there.

**Comment by jhalterman** [comment]
To simplify the discussion, I'm going to close this as a duplicate of #208 and take the conversation over there. I might like a PR for this, but let's wait till that discussion settles first.

**Comment by jhalterman** [comment]
This was released in 2.4.0. Check the docs:

- https://jodah.net/failsafe/circuit-breaker/#opening
- https://jodah.net/failsafe/circuit-breaker/#time-based-resolution
